activation/activation.go

```diff
@@ -383,60 +324,57 @@ func (b *Builder) loadChallenge() error {
 	return nil
 }
 
-func (b *Builder) discardChallenge() error {
-	return b.store.Put(b.getNipstKey(), []byte{})
-}
-
-// PublishActivationTx attempts to publish an atx for the given epoch, it returns an error if an atx cannot be created
-// publish atx may not produce an atx each time it is called, that is expected behaviour as well.
-func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
-	if b.nipst != nil {
-		b.log.With().Info("re-entering atx creation in epoch %v", log.EpochId(uint64(epoch)))
+// PublishActivationTx attempts to publish an atx, it returns an error if an atx cannot be created.
+func (b *Builder) PublishActivationTx() error {
+	b.refreshChallenge()
+	if b.challenge != nil {
+		b.log.With().Info("using existing challenge", log.EpochId(uint64(b.currentEpoch())))
 	} else {
-		b.log.With().Info("starting build atx in epoch %v", log.EpochId(uint64(epoch)))
-		if b.challenge != nil {
-			if b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch) < epoch {
-				log.Info("previous challenge loaded from store, but epoch has already passed, %s, %s", epoch, b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch))
-				b.challenge = nil
-				err := b.discardChallenge()
-				if err != nil {
-					log.Error("cannot discard challenge")
-				}
-			}
-		}
-		if b.challenge == nil {
-			err := b.buildNipstChallenge(epoch)
-			if err != nil {
-				if _, alreadyPublished := err.(alreadyPublishedErr); alreadyPublished {
-					events.Publish(events.AtxCreated{false, "", uint64(epoch)})
-					return nil
-				}
-				return err
-			}
-		}
-		hash, err := b.challenge.Hash()
+		b.log.With().Info("building new challenge", log.EpochId(uint64(b.currentEpoch())))
+		err := b.buildNipstChallenge()
 		if err != nil {
-			return fmt.Errorf("getting challenge hash failed: %v", err)
+			return err
 		}
-		b.nipst, err = b.nipstBuilder.BuildNIPST(hash)
-		if err != nil {
-			return fmt.Errorf("cannot create Nipst: %v", err)
+	}
+	pubEpoch := b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch)
+
+	hash, err := b.challenge.Hash()
+	if err != nil {
+		return fmt.Errorf("getting challenge hash failed: %v", err)
+	}
+	// ⏳ the following method waits for a PoET proof, which should take ~1 epoch
+	nipst, err := b.nipstBuilder.BuildNIPST(hash, b.layerClock.AwaitLayer((pubEpoch + 2).FirstLayer(b.layersPerEpoch)), b.stop)
```

Comment:
why pass the channel and not the pubEpoch, setting timer inside BuildNIPST func?

Reply:
I disagree. The NISPT builder has no concept of time or layers, why involve it in this decision? It shouldn't be aware of the fact that an ATX is valid until the end of it's target epoch, or that the target epoch is one after the publication epoch. These are ATX builder concerns. It calls the NIPST builder and tells it "build me a NIPST before this time or return an error" which makes perfect sense to me.

I'm actually creating this same timeout in another place, so I think I'll extract it and call it `atxExpired` instead of timeout which sounds arbitrary.